### DISCUSSION TITLE
How to change the default toolchain?

### DISCUSSION THREAD

**Discussion by nikhilweee** [astral-sh/rye#1276] [Q&A] [answered]
I am trying to update the default toolchain used by rye. Note that this is not the project specific toolchain.

```console
$ cat ~/.rye/config.toml 
[behavior]
use-uv = true

[default]
toolchain = "cpython@3.12.4"
```

```console
$ rye toolchain remove 3.12.2
error: toolchain cpython@3.12.2 is still in use by rye itself
```

```console
$ rye self update
Updating to latest
Checking checksum
Validate updated installation
Initializing new virtualenv in /tmp/.tmppOk3zx/.venv
Python version: cpython@3.12.4
Updated!

rye 0.37.0
commit: 0.37.0 (09b67c469 2024-07-20)
platform: linux (x86_64)
self-python: cpython@3.12.2
symlink support: true
uv enabled: true
```

How can I make rye use `3.12.4` instead of `3.12.2`? I already updated the default toolchain in `config.toml` but that doesn't help.

**Comment by philippemiron** [comment]
You can see all the available versions with `rye toolchain list --include-downloadable`. If you change the version's number in `.python-version` and do `rye sync`, it should update everything.

**Reply by nikhilweee** [reply]
> I am trying to update the default toolchain used by rye. Note that this is not the project specific toolchain.

As I said before, I am talking about the default toolchain used by rye, not the project specific toolchain.

**Reply by philippemiron** [reply]
Sorry, I read too fast. I'm also confused by the `default.toolchain` now. I switched to `3.10` to test it out, but new projects are still initialized with `3.12.4`.

**Comment by bluss** [comment, marked as answer]
Rye updates its self-venv if it knows its version marker for that venv has changed (happens in some version updates of Rye but not all). It will use the newest available toolchain to do so.

With some knowledge of Rye internals you can do this:

```bash
rye fetch 3.12.4 # (ensure it's available).
rm .rye/self/tool-version.txt   # remove or change the version marker
```

Then do anything where rye wants to have its self venv. For example sync in one of your projects:

```
rye sync
```
```
Detected outdated rye internals. Refreshing
Bootstrapping rye internals
```

**Reply by nikhilweee** [reply]
Just to add another example, one can also do `rye self update` to trigger a refresh if they are not in an existing project.
